include/matx/transforms/cub.h

```diff
@@ -1092,6 +1096,163 @@ inline void ExecSort(OutputTensor &a_out,
+    CubParams_t params;
+
+    for (int r = 0; r < InputOperator::Rank(); r++) {
+      params.size.push_back(a.Size(r));
+    }
+
+    params.op = op;
+    if constexpr (OutputTensor::Rank() > 0)
+    {
+      params.batches = TotalSize(a_out);
+    }
+    else
+    {
+      params.batches = 1;
+    }
+    params.dtype = TypeToInt<T1>();
+    params.stream = stream;
+
+    return params;
+  }
+
+  /**
+   * Execute an argmax on a tensor
+   *
+   * @note Views being passed must be in row-major order
+   *
+   * @tparam OutputTensor
+   *   Type of output tensor
+   * @tparam TensorIndexType
+   *   Type of the output index tensor
+   * @tparam InputOperator
+   *   Type of input tensor
+   * @param a_out
+   *   Output tensor
+   * @param aidx_out
+   *   Output index tensor
+   * @param a
+   *   Input tensor
+   * @param stream
+   *   CUDA stream
+   *
+   */
+  inline void ExecArgMax(OutputTensor &a_out,
+                         TensorIndexType &aidx_out,
+                         const InputOperator &a,
+                         const cudaStream_t stream)
+  {
+#ifdef __CUDACC__
+    MATX_NVTX_START("", matx::MATX_NVTX_LOG_INTERNAL)
+
+    CustomArgMaxCmp cmp_op;
+    auto initial_value = cuda::std::tuple<matx::index_t,typename InputOperator::value_type>(-1, std::numeric_limits<typename InputOperator::value_type>::lowest());
+
+    if constexpr (OutputTensor::Rank() > 0) {
+      auto a_iter = matx::RandomOperatorOutputIterator{a};
+      auto zipped_input = thrust::make_zip_iterator(thrust::make_counting_iterator<matx::index_t>(0), a_iter);
+      auto zipped_output = thrust::make_zip_iterator(aidx_out.Data(), a_out.Data());
+
+      int BATCHES = static_cast<int>(TotalSize(a_out));
+      int N = static_cast<int>(TotalSize(a)) / BATCHES;
+
+      auto r0 = matx::range<0>({BATCHES},0,N);
+      auto r0_iter = matx::RandomOperatorIterator{r0};
+      auto r1 = matx::range<0>({BATCHES},N,N);
+      auto r1_iter = matx::RandomOperatorIterator{r1};
+
+      cub::DeviceSegmentedReduce::Reduce(
+        d_temp,
+        temp_storage_bytes,
+        zipped_input,
+        zipped_output,
+        BATCHES,
+        r0_iter,
+        r1_iter,
+        cmp_op,
+        initial_value,
+        stream);
+    }
+    else {
+      auto a_iter = matx::RandomOperatorOutputIterator{a};
+      auto zipped_input = thrust::make_zip_iterator(thrust::make_counting_iterator<matx::index_t>(0), a_iter);
+      auto zipped_output = thrust::make_zip_iterator(aidx_out.Data(), a_out.Data());
+
+      int N = static_cast<int>(TotalSize(a));
+
+      cub::DeviceReduce::Reduce(
+        d_temp,
+        temp_storage_bytes,
+        zipped_input,
+        zipped_output,
+        N,
+        cmp_op,
+        initial_value,
+        stream);
+    }
+#endif
+  }
+
+  /**
+   * Destructor
+   *
+   * Destroys any helper data used for provider type and any workspace memory
+   * created
+   *
+   */
+  ~matxCubTwoOutputPlan_t()
+  {
+    matxFree(d_temp, cudaStreamDefault);
```

Comment:
Why is it using `cudaStreamDefault` if it was allocated in a stream above?

Reply:
Copy and paste from matxCubOutputPlan_t destructor

Reply:
Thanks, that's probably a hack to get around the static destruction. I'll see if we can remove it.